runtime/jvm/src/test/kotlin/kotlinx/serialization/formats/json/JsonTreeAndMapperTest.kt

```diff
@@ -0,0 +1,83 @@
+package kotlinx.serialization.formats.json
+
+import kotlinx.serialization.*
+import kotlinx.serialization.json.*
+import kotlin.test.Test
+import kotlin.test.assertEquals
+
+@Serializable
+data class Payload(val from: Long, val to: Long, val msg: String)
+
+sealed class DummyEither {
+    data class Left(val errorMsg: String): DummyEither()
+    data class Right(val data: Payload): DummyEither()
+}
+
+@Serializer(forClass = DummyEither::class)
+object EitherSerializer: KSerializer<DummyEither> {
+    override fun load(input: KInput): DummyEither {
+        val jsonReader = input as? JSON.JsonInput
+                ?: throw SerializationException("This class can be loaded only by JSON")
+        val tree = jsonReader.readAsTree() as? JsonObject
+                ?: throw SerializationException("Expected JSON object")
+        if ("error" in tree) return DummyEither.Left(tree.getAsValue("error").str)
+        return DummyEither.Right(JsonTreeMapper().readTree(tree, Payload.serializer()))
+    }
+
+    override fun save(output: KOutput, obj: DummyEither) {
+        val tree = when (obj) {
+            is DummyEither.Left -> JsonObject(mapOf("error" to JsonString(obj.errorMsg)))
+            is DummyEither.Right -> JsonTreeMapper().writeTree(obj.data, Payload.serializer())
+        }
+        val jsonWriter = output as? JSON.JsonOutput
```

Comment:
Maybe `jsonWrite` could go to the top of this method to fail fast and mirror that `jsonReader` is also on the top in `fun load()`.